ethcore/src/state/mod.rs

```diff
@@ -855,16 +855,29 @@ impl<B: Backend> State<B> {
 		}))
 	}
 
-	fn query_pod(&mut self, query: &PodState) -> trie::Result<()> {
-		for (address, pod_account) in query.get() {
+	// Return a list of all touched addresses in cache.
+	fn touched_addresses(&self) -> Vec<Address> {
+		assert!(self.checkpoints.borrow().is_empty());
+		self.cache.borrow().iter().map(|(add, _)| *add).collect()
+	}
+
+	fn query_pod(&mut self, query: &PodState, touched_addresses: &[Address]) -> trie::Result<()> {
+		let pod = query.get();
+
+		for address in touched_addresses {
```

Comment:
Can you explain how `filter_map` would work in this case? Here we need to call `ensure_cached` for all items in `touched_addresses` first, so I think it might not save us LOC or code clarity. Do you mean something like this?

```
let cached_pod_accounts = touched_addresses.filter_map(|address| {
  if !self.ensure_cached(address, RequireCache::Code, true, |a| a.is_some())? {
    None
  } else {
    pod.get(address)
  }
});

for pod_account in cached_pod_accounts {
  for key in pod_account.storage.keys() {
    self.storage_at(address, key)?;
  }
}
```

Reply:
Oh, I think I get it now. Sorry for the confusion, had to go through the code to understand what's going on.

So if I get that correctly:
1. For every `touched_address` we need to call `ensure_cached` to populate the pre-state cache.
2. Additionally for every `touched_address` that has an account in pre-state we need to query all storage keys (from the post-state).

The current loop is the most clear way to do it then.

That said, can we get a test for it?

Reply:
I added `should_trace_diff_suicided_accounts` that will test `touched_addresses` and `State::diff_from`.